sympy/polys/rootoftools.py

```diff
@@ -772,15 +775,94 @@ def _get_roots(cls, method, poly, radicals):
         free_names = {str(i) for i in poly.free_symbols}
         for x in chain((symbols('x'),), numbered_symbols('x')):
             if x.name not in free_names:
+                # xreplace makes alg field domain back into EX
+                # XXX: so make sure it stays same?
                 poly = poly.xreplace({d: x})
+                if dom.is_AlgebraicField:
+                    poly = Poly(poly, domain=dom)
                 break
+
+        if dom.is_QQ or dom.is_ZZ:
+            return cls._get_roots_qq(method, poly, radicals)
+            # XXX: any better way to check if real?
+            # without this check, some doctests fail (with complex terms)
+        if dom.is_AlgebraicField and all(c.is_real for c in poly.coeffs()):
+            return cls._get_roots_alg(method, poly, radicals)
+        else:
+            # XXX: not sure how to handle ZZ[x] which appears in some tests?
+            # this makes the tests pass alright but has to be a better way?
+            return cls._get_roots_qq(method, poly, radicals)
+
+
+    @classmethod
+    def _get_roots_qq(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with rational coefficients. """
         coeff, poly = cls._preprocess_roots(poly)
         roots = []
 
         for root in getattr(cls, method)(poly):
             roots.append(coeff*cls._postprocess_root(root, radicals))
+
         return roots
 
+    @classmethod
+    def _get_roots_alg(cls, method, poly, radicals):
+        """Return postprocessed roots of specified kind
+         for polynomials with algebraic coefficients It assumes
+         the domain is already an algebraic field. """
+
+        poly_lift = poly.lift()
+
+        coeff, poly_lift = cls._preprocess_roots(poly_lift)
+        roots = []
+
+        for root in getattr(cls, method)(poly_lift):
+            roots.append(coeff*cls._postprocess_root(root, radicals))
+
+        if method == "_real_roots":
+            roots = cls._numerically_filter_roots(poly, roots, real=True)
+        elif method == "_all_roots":
+            roots = cls._numerically_filter_roots(poly, roots, real=False)
+
+        return roots
+
+    @classmethod
+    def _numerically_filter_roots(cls, poly, candidates, real):
+        # must be QQ, ZZ, or Alg for proper counting
+        dom = poly.get_domain()
+        if not (dom.is_ZZ or dom.is_QQ or dom.is_AlgebraicField):
+            raise NotImplementedError(
+                "root counting not supported over %s" % dom)
+
+        if real:
+            num_roots = poly.count_roots()
+        else:
+            num_roots = poly.degree()
+
+        prec = 10
+        # compare len(set()) bc expected behavior is for multiple roots
+        while len(set(candidates)) > num_roots:
```

Comment:
Is the multiple root case not handled earlier somehow?

I would have thought that by the time we get here the polynomial is irreducible and so cannot have multiple roots.

Reply:
Before the point `_numerically_filter_roots()` is called, the roots are created by `ComplexRootOf._real_roots()` or `ComplexRootOf._all_roots()` which return multiples.

Eg
```
>>> ComplexRootOf._real_roots(Poly((x-1)**2))
[(PurePoly(x - 1, x, domain='ZZ'), 0), (PurePoly(x - 1, x, domain='ZZ'), 0)]
```

Reply:
That makes me think that we might be handling this in the wrong place. The RootOf code is a bit convoluted which makes it hard to follow through.

Potentially it is better to compute the square free factorisation of the polynomial first and then do the lift on the square free parts. Then we don't have multiple roots and we work with smaller polynomials.

I might be wrong though and maybe it is just best to do the lift at the outset so that all the other steps can use efficient calculations in `QQ`.

Reply:
My understanding of the RootOf code is (eg for real roots):

`_get_roots()` --> `_real_roots()` --> gets factors, calls `_get_reals()` --> `_get_reals_sqf()` which computes root isolating intervals for a square free factor

So it seems like the square free stuff is done further along currently. If doing the lift is better on smaller polynomials, then we could just change `_get_roots_alg()` to incorporate square free factorization before calling `_get_roots()`.

Reply:
> If doing the lift is better on smaller polynomials, then we could just change `_get_roots_alg()` to incorporate square free factorization before calling `_real_roots()`.

I'm not sure that it is better to do this. Factorising over extension fields is also expensive so it might actually be better to lift the original polynomial so that all subsequent calculations are done with only polynomials having rational coefficients. Hard to say without more analysis/timings. For now I think this is fine if it works.